lib/stm32/g0/adc.c

```diff
@@ -0,0 +1,251 @@
+			reg32 &= ~(ADC_SMPR_SMP2_MASK << ADC_SMPR_SMP2_SHIFT);
+			reg32 |= (time << ADC_SMPR_SMP2_SHIFT);
+			break;
+	}
+	ADC_SMPR1(adc) = reg32;
+}
+
+/** @brief ADC Set a Regular Channel Conversion Sequence
+ *
+ * Define a simple sequence of channels to be converted.
+ * ADCSTART must be de-asserted before sequence setup.
+ *
+ * @param[in] adc ADC base address (@ref adc_reg_base)
+ * @param[in] length Number of channels in the group, range 0..18
+ * @param[in] channel Set of channels in sequence (0..18 or @ref adc_channel)
+ */
+void adc_set_regular_sequence(uint32_t adc, uint8_t length, uint8_t channel[])
+{
+	uint32_t reg32 = 0;
+	bool stepup = false, stepdn = false;
+
+	if (length > ADC_CHSELR_MAX_CHANNELS) {
+		return;
+	}
+
+	if (length == 0) {
+		ADC_CHSELR(adc) = 0;
+		return;
+	}
+
+	reg32 |= (1 << channel[0]);
+
+	for (uint8_t i = 1; i < length; i++) {
+		reg32 |= ADC_CHSELR_CHSEL(channel[i]);
+		stepup |= channel[i-1] < channel[i];
+		stepdn |= channel[i-1] > channel[i];
+	}
+
+	/* Check if the channel list is in order */
+	if (stepup && stepdn) {
+		cm3_assert_not_reached();
+	}
+
+	/* Each modification to ADC_CFGR1's SCANDIR or CHSELRMOD bits or
+	   ADC_CHSELR register must be done after previous configuration change
+	   being properly applied: We have to clear ccrdy bit before and poll for
+	   it being assert back after, before going on. We also need to wait for
+	   configuration applied before starting conversion, or start will be
+	   ignored.  */
+
+	/* Setup scandir, if needed, waiting for configuration be applied.. */
+	if (stepdn && (!(ADC_CFGR1(adc) & ADC_CFGR1_SCANDIR))) {
+		ADC_ISR(adc) &= ~ADC_ISR_CCRDY;
+		ADC_CFGR1(adc) |= ADC_CFGR1_SCANDIR;
+		while (!(ADC_ISR(adc) & ADC_ISR_CCRDY));
+	} else if (stepup && ((ADC_CFGR1(adc) & ADC_CFGR1_SCANDIR))) {
+		ADC_ISR(adc) &= ~ADC_ISR_CCRDY;
+		ADC_CFGR1(adc) &= ~ADC_CFGR1_SCANDIR;
+		while (!(ADC_ISR(adc) & ADC_ISR_CCRDY));
+	}
+
+	/* Setup ADC in simple, not configurable, mode, if needed. */
+	if ((ADC_CFGR1(adc) & ADC_CFGR1_CHSELRMOD)) {
+		ADC_ISR(adc) &= ~ADC_ISR_CCRDY;
+		ADC_CFGR1(adc) &= ~ADC_CFGR1_CHSELRMOD;
+		while (!(ADC_ISR(adc) & ADC_ISR_CCRDY));
+	}
+
+	if (ADC_CHSELR(adc) != reg32) {
+		ADC_ISR(adc) &= ~ADC_ISR_CCRDY;
+		ADC_CHSELR(adc) = reg32;
+		while (!(ADC_ISR(adc) & ADC_ISR_CCRDY));
+	}
+}
+
+/** @brief ADC Set an Injected Channel Conversion Sequence
+ *
+ * Setup the ADC in fully configurable sequence mode, and set a user defined
+ * sequence of channels to be converted. ADCSTART must be de-asserted before
+ * sequence setup.
+ * Note that on STM32G0, the sequence is limited to 8 channels, and
+ * only the first 14 ADC channels are selectable.
+ *
+ * @param[in] adc ADC base address (@ref adc_reg_base)
+ * @param[in] length Number of channels in the group, range 0..8
+ * @param[in] channel Set of channels in sequence (0..14 or @ref adc_channel)
+ */
+void adc_set_injected_sequence(uint32_t adc, uint8_t length, uint8_t channel[])
```

Comment:
I understand why you went to use this api for the "other" way of doing sequences on g0, (just the first 14, and only 8 channels) but the "injected" name has a very specific meaning on the other families, and I think it's misleading to try and use that name here.

I don't know what _to_ do, but these two sequence limitions are quite awkward

Reply:
hmm, It looked like the f3 injected stuff - i'll have a look at the refman later.
I pushed a patch without the "injected" stuff meanwhile, to merge the remaining stuff.